music/src/ddsp/model.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * Core implementation for [DDSP]{@link
+ * https://g.co/magenta/ddsp} models.
+ *
+ * @license
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Imports.
+ */
+import * as tf from '@tensorflow/tfjs';
+import '@tensorflow/tfjs-backend-webgl';
+
+import { getAudioFeatures, toneTransfer, memCheck } from './ddsp';
+import { MODEL } from './constants';
+import { AudioFeatures } from './interfaces';
+
+class DDSP {
+  private initialized: boolean;
+
+  /**
+   * `DDSP` constructor.
+   *
+   * @param model The model to use
+   */
+  constructor() {
+    this.initialize();
+  }
+
+  /**
+   * Loads variables from the checkpoint and builds the model graph.
+   */
+  async initialize() {
+    tf.registerOp('Roll', (node) => {
+      const tensors = tf.split(node.inputs[0], 2, 2);
+      const result = tf.concat([tensors[1], tensors[0]], 2);
+      tensors.forEach((tensor) => tensor.dispose());
+      return result;
+    });
+    tf.env().set('WEBGL_PACK', false);
+    tf.env().set('WEBGL_CONV_IM2COL', false);
+    tf.env().set('WEBGL_DELETE_TEXTURE_THRESHOLD', 100 * 1024 * 1024);
+  }
+
+  /**
+   * Disposes of any untracked `Tensors` to avoid GPU memory leaks.
+   */
+  dispose() {
+    if (!this.initialized) {
+      return;
+    }
+    this.initialized = false;
+  }
+
+  /**
+   * Returns true if model is intialized.
+   */
+  isInitialized() {
+    return this.initialized;
+  }
+
+  /**
+   * Include memCheck as a member method for API/export.
+   *
+   * @param memCheck An AudioBuffer
+   * @returns Audio Features of provided audio buffer
+   */
+  async memCheck(): Promise<boolean> {
+    return await memCheck();
+  }
+
+  /**
+   * Include getAudioFeatures as a member method for API/export.
+   *
+   * @param inputAudioBuffer An AudioBuffer
+   * @returns Audio Features of provided audio buffer
+   */
+  async getAudioFeatures(
+    inputAudioBuffer: AudioBuffer
+  ): Promise<AudioFeatures> {
+    return await getAudioFeatures(inputAudioBuffer);
+  }
+
+  /**
+   * Synthesizes audio features based on model.
+   *
+   * @param inputAudioBuffer An AudioBuffer
+   * @returns Audio Features of provided audio buffer
+   */
+  async synthesize(
+    model: MODEL,
+    audioFeatures: AudioFeatures
+  ): Promise<Float32Array> {
+    return await toneTransfer(model, audioFeatures);
```

Comment:
not to be picky about this, but technically toneTransfer() is the process of both extracting the audio features and then synthesizing.

You could rename toneTransfer to synthesize, and then have another convenience function toneTransfer that just takes in audio and returns the tone transferred audio. 

Also, I'm not sure it makes sense to have the synthesize function (toneTransfer() at the moment) as a separate method, because it always needs to be paired with a model, so it might make sense to just have it defined in the class itself.

Reply:
I thought a little bit about this and decided to do it this way because of "getAudioFeatures" is actually agnostic of the models, and if you see the demo, you can see that you just need to run getAudioFeatures once, and then run synthesize with all 4 different models. WDYT?

Reply:
That's fine, but then the method should be called `synthesize` because that's what it does :). I'm just saying if someone makes a demo with only one model and wants a convenience function that goes straight audio2audio then a convenience function named toneTransfer would make sense.